src/Common/Exception.cpp

```diff
@@ -63,11 +64,18 @@ void handle_error_code([[maybe_unused]] const std::string & msg, int code, bool
     ErrorCodes::increment(code, remote, msg, trace);
 }
 
-Exception::Exception(const std::string & msg, int code, bool remote_)
-    : Poco::Exception(msg, code)
+Exception::MessageMasked::MessageMasked(const std::string & msg_)
+    : msg(msg_)
+{
+    if (auto * masker = SensitiveDataMasker::getInstance())
+        masker->wipeSensitiveData(msg);
+}
+
+Exception::Exception(const MessageMasked & msg_masked, int code, bool remote_)
+    : Poco::Exception(msg_masked.msg, code)
     , remote(remote_)
 {
-    handle_error_code(msg, code, remote, getStackFramePointers());
+    handle_error_code(msg_masked.msg, code, remote, getStackFramePointers());
```

Comment:
I'm not sure if it can really be done 100% safe on AST level... 

Imagine you can't parse the query correctly, but it have sensitive data, or that sensitive data does not come from the query at all. So i think that those 'sensitivedatamaster' is still needed (and also it's ok to have some predefined stuff on AST level)

And test - sure you can reuse it.

Reply:
I didn't suggest removing the regular expression approach. In my opinion it's better to use both: first hardcoded wiping based on AST and then customizable wiping based on regexp.